### PR TITLE
Method and tests for jep334 MethodHandle

### PR DESCRIPTION
**Pull request by theresa-m**
- describeConstable

Depends on (see https://github.com/eclipse/openj9/issues/4195): 
- Fix Java 12 build failures: https://github.com/eclipse/openj9/pull/3980
- Constants API stubs: https://github.com/eclipse/openj9/pull/4030
- Java12andUp test setup: https://github.com/eclipse/openj9/pull/4104
- Jep 334 Class methods: https://github.com/eclipse/openj9/pull/4157
- Jep 334 MethodType methods: https://github.com/eclipse/openj9/pull/4194

Signed-off-by: Theresa Mammarella <Theresa.T.Mammarella@ibm.com>

### CODE REVIEW

**Comment by pshipton** [comment]
MethodHandle implements Constable
http://cr.openjdk.java.net/~iris/se/12/build/latest/api/java.base/java/lang/invoke/MethodHandle.html

MethodType implements Constable, TypeDescriptor.OfMethod<Class<?>,MethodType>
http://cr.openjdk.java.net/~iris/se/12/build/latest/api/java.base/java/lang/invoke/MethodType.html

VarHandle and VarHandle.AccessMode implements Constable
http://cr.openjdk.java.net/~iris/se/12/build/latest/api/java.base/java/lang/invoke/VarHandle.html
http://cr.openjdk.java.net/~iris/se/12/build/latest/api/java.base/java/lang/invoke/VarHandle.AccessMode.html

File: test/functional/Java12andUp/src/org/openj9/test/java_lang_invoke/helpers/Helper.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.MethodHandles.Lookup;
+import java.lang.invoke.VarHandle;
+
+public class Helper implements HelperInterface {
+    /*
+     * MethodHandle helper fields & methods
+     */
+    /* fields */
+    public volatile int getterTest = 1;
```
Comment (DanHeidinga) [inline test/functional/Java12andUp/src/org/openj9/test/java_lang_invoke/helpers/Helper.java]: formatting

File: test/functional/Java12andUp/src/org/openj9/test/java_lang_invoke/helpers/Helper.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+package org.openj9.test.java_lang_invoke.helpers;
+
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.MethodHandles.Lookup;
+import java.lang.invoke.VarHandle;
+
+public class Helper implements HelperInterface {
```
Comment (DanHeidinga) [inline test/functional/Java12andUp/src/org/openj9/test/java_lang_invoke/helpers/Helper.java]: Can a more explicit name be chosen?  `Helper` could be a lot of different things

File: test/functional/Java12andUp/src/org/openj9/test/java_lang_invoke/helpers/HelperInterface.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+package org.openj9.test.java_lang_invoke.helpers;
+
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.MethodHandles.Lookup;
+
+/* methods to test MethodHandle interface types */
+public interface HelperInterface {
```
Comment (DanHeidinga) [inline test/functional/Java12andUp/src/org/openj9/test/java_lang_invoke/helpers/HelperInterface.java]: Can a more specific name be chosen?

File: test/functional/Java12andUp/testng.xml [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+  SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+-->
+
+<!DOCTYPE suite SYSTEM "http://testng.org/testng-1.0.dtd">
+<suite name="Java12andUp suite" parallel="none" verbose="2">
+
+	<test name="Jep334Tests">
+		<classes>
+      <class name="org.openj9.test.java_lang.Test_String" />
```
Comment (DanHeidinga) [inline test/functional/Java12andUp/testng.xml]: formatting looks off here

**Comment by DanHeidinga** [comment]
I think there's a simpler approach for this - use `java.lang.invoke.MethodHandleInfo` as it already can describe all "crackable" MethodHandles.  Rather than spreading the logic to create a right kind of Constable across the MH subclasses, it can be centralized in `MethodHandle` and just needs to know how to convert an MHInfo into a ConstantDesc.

The MHInfo can be looked up using https://github.com/eclipse/openj9/blob/7ce88fd12da2001f380477ea832c71d15a207fd3/jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java#L1529

We can use this special lookup to bypass the security checks when cracking an MH:
https://github.com/eclipse/openj9/blob/7ce88fd12da2001f380477ea832c71d15a207fd3/jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java#L179 

which should hopefully centralize the code and simplify the implementation as all non-crackable MHs can just return an empty Optional.

File: test/functional/Java12andUp/src/org/openj9/test/java_lang_invoke/Test_MethodHandle.java [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+			try {
+				resolvedHandle = (MethodHandle)handleDesc.resolveConstantDesc(resolveLookup);
+			} catch(ReflectiveOperationException e) {
+				Assert.fail(testName + ": ReflectiveOperationException thrown at MethodHandleDesc.resolveConstantDesc() call");
+				return;
+			}
+
+			String originalString = handle.type().descriptorString();
+			String newString = resolvedHandle.type().descriptorString();
+			logger.debug(testName + ": Descriptor of original MethodHandle " + originalString + " descriptor of MethodHandleDesc is: " + newString);
```
Comment (DanHeidinga) [inline test/functional/Java12andUp/src/org/openj9/test/java_lang_invoke/Test_MethodHandle.java]: We don't need to convert to Strings here - the MethodType objects can be directly compared.

File: test/functional/Java12andUp/src/org/openj9/test/java_lang_invoke/Test_MethodHandle.java [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
+import java.util.Optional;
+
+/**
+ * This test Java.lang.invoke.MethodHandle API added in Java 12 and later version.
+ * - describeConstable: one test for each constant in DirectMethodHandleDesc.Kind to make sure each type is supported
+ *
+ */
+public class Test_MethodHandle {
```
Comment (DanHeidinga) [inline test/functional/Java12andUp/src/org/openj9/test/java_lang_invoke/Test_MethodHandle.java]: MethodHandleInfo may also help with the testing as we can validate that the MHInfo and the ConstantDesc describe the same info.

**Comment by theresa-m** [comment]
@DanHeidinga I refactored describeConstable so everything is done in MethodHandle using MethodHandleInfo to help.

**Comment by theresa-m** [comment]
ready for another review

**Comment by DanHeidinga** [comment]
@theresa-m Can you rebase and resolve the conflict?

**Comment by theresa-m** [comment]
its resolved now. note this or is dependent on https://github.com/eclipse/openj9/pull/4194

**Comment by DanHeidinga** [comment]
Jenkins test sanity xlinux jdk12

**Comment by DanHeidinga** [comment]
I've addressed the merge conflict and kicked off a PR build.  Will check this again Saturday and hopefully merge it then

**Pull request merged**